runtime/domain.c

```diff
@@ -1713,7 +1713,17 @@ static void domain_terminate (void)
       caml_plat_broadcast(&s->cond);
       caml_plat_unlock(&s->lock);
 
+      /* signal the domain termination to the backup thread
+         NB: for a program with no additional domains, the backup thread
+         will not have been started */
+      atomic_store_rel(&domain_self->backup_thread_msg, BT_TERMINATE);
+      caml_plat_signal(&domain_self->domain_cond);
+      caml_plat_unlock(&domain_self->domain_lock);
+
       CAMLassert (domain_self->backup_thread_running);
+      /* CR gasche: why is this assertion correct in the case where
```

Comment:
Isn't it because the main OCaml thread does not go through `domain_terminate`?

Reply:
That would be my guess as well, thanks! But I understand that this could change in the future. Should we remove the assertion? (Or maybe this could or should become an `if (domain_self->backup_thread_running) { ... }`, disabling also the BT_TERMINATE signalling.) In any case, it is currently weird that the comment that is now shortly above contradicts this assertion.

(I wonder what happens if we start calling `domain_terminate` for the main domain as well. Do other things break, or is it supposed to work?)

Reply:
I'm not sure this assertion is useful anymore, I'm in favor of removing it at this point.

If a thread goes through `domain_terminate`, then a backup thread _must_ be running, and it feels weird to check for this invariant at termination.

For Domain 0 calling _domain_terminate_, I am not sure this is meant to happen.  Domain 0 returning means all remaining threads will be shut down on exit.

However, I see issues coming when we re-implement proper cleanup at exit, where this dual path of cleaning and exiting may come back to bite us.

(and, more, while not quite relevant, a single domain 0 does not mean it has no backup thread: it may very well be that all over domains have shut down, with domain 0 retaining its backup thread after.)